framework/src/auxkernels/AuxKernel.C

```diff
@@ -138,6 +140,30 @@ AuxKernelTempl<ComputeValueType>::AuxKernelTempl(const InputParameters & paramet
   for (const auto & it : coupled_vars)
     for (const auto & var : it.second)
       _depend_vars.insert(var->name());
+
+  if (_boundary_restricted && !isNodal())
+  {
+    // when the variable is elemental and this aux kernel operates on boundaries,
+    // we need to check no elements are visited more than once through visiting
+    // all the sides on the boundaries
+    auto boundaries = _mesh.getMesh().get_boundary_info().build_side_list();
+    std::set<dof_id_type> elements;
+    for (const auto & t : boundaries)
+    {
+      if (hasBoundary(std::get<2>(t)))
+      {
+        auto eid = std::get<0>(t);
+        std::pair<std::set<dof_id_type>::iterator, bool> stat = elements.insert(eid);
+        if (!stat.second)
+          mooseError("Boundary restricted auxiliary kernel ",
+                     name(),
+                     " has element (ID=",
+                     eid,
+                     ") connected with more than one boundary sides.\nYou can consider using "
+                     "multiple auxilary variables/kernels each on a part of the boundaries.");
```

Comment:
```suggestion
                     ") that has multiple boundary sides.\nSee the AuxKernel documentation on boundary restricted aux kernels for understanding this error.");
```

Reply:
We're looking into being able to link to documentation from within source... would be nice